File: scripts/fix-syntax-errors.ts
#!/usr/bin/env tsx

/**
 * Script to fix remaining syntax errors from doubled conditionals
 */

import * as fs from 'fs';
import * as path from 'path';
import { glob } from 'glob';

const SRC_DIR = path.join(process.cwd(), 'src');

async function fixSyntaxErrors() {
  console.log('🔍 Fixing syntax errors from doubled conditionals...\n');

  const files = await glob('**/*.{ts,tsx}', {
    cwd: SRC_DIR,
    ignore: ['**/node_modules/**'],
  });

  let fixedCount = 0;

  for (const file of files) {
    const filePath = path.join(SRC_DIR, file);
    let content = fs.readFileSync(filePath, 'utf8');
    let modified = false;

    // Pattern 1: Fix doubled conditional in single expression
    // error instanceof Error ? error.message : String(error) : String(error)
    // Should be: error instanceof Error ? error.message : String(error)
    const doubledConditionalPattern = /(\w+)\s+instanceof\s+Error\s*\?\s*(\1\.\w+)\s*:\s*String\(\1\)\s*:\s*[^,;}\n]+/g;
    
    if (doubledConditionalPattern.test(content)) {
      content = content.replace(doubledConditionalPattern, (match, varName, property) => {
        // Extract just the first conditional
        const mainConditional = `${varName} instanceof Error ? ${property} : String(${varName})`;
        return mainConditional;
      });
      modified = true;
    }

    // Pattern 2: Fix nested conditionals in object properties
    // errorMessage: error instanceof Error ? error.message : String(error) : String(error)
    const objectPropertyPattern = /(\w+):\s*(\w+)\s+instanceof\s+Error\s*\?\s*(\2\.\w+)\s*:\s*String\(\2\)\s*:\s*[^,}\n]+/g;
    
    if (objectPropertyPattern.test(content)) {
      content = content.replace(objectPropertyPattern, (match, prop, varName, property) => {
        return `${prop}: ${varName} instanceof Error ? ${property} : String(${varName})`;
      });
      modified = true;
    }

    // Pattern 3: Fix stacked conditionals
    // error instanceof Error ? error instanceof Error ? error.message : String(error) : other
    const stackedConditionalPattern = /(\w+)\s+instanceof\s+Error\s*\?\s*\1\s+instanceof\s+Error\s*\?\s*(\1\.\w+)\s*:\s*String\(\1\)\s*:\s*([^,;}\n]+)/g;
    
    if (stackedConditionalPattern.test(content)) {
      content = content.replace(stackedConditionalPattern, (match, varName, property, fallback) => {
        return `${varName} instanceof Error ? ${property} : ${fallback}`;
      });
      modified = true;
    }

    // Pattern 4: Fix malformed error message extractions
    // .error instanceof Error ? updateSettings.error.message : String(updateSettings.error)
    // Where it was incorrectly transformed
    const malformedPropertyAccess = /\.\((\w+)\s+instanceof\s+Error\s*\?\s*\1\.\w+\s*:\s*String\(\1\)\)/g;
    
    if (malformedPropertyAccess.test(content)) {
      content = content.replace(malformedPropertyAccess, (match, varName) => {
        return `.error instanceof Error ? ${varName}.error.message : String(${varName}.error)`;
      });
      modified = true;
    }

    // Pattern 5: Clean up errorMsg{number} variables that reference undefined 'error'
    // const errorMsg123 = error instanceof Error ? error.message : String(error);
    // Where 'error' doesn't exist in scope
    const lines = content.split('\n');
    for (let i = 0; i < lines.length; i++) {
      const line = lines[i];
      if (/const errorMsg\d+\s*=/.test(line)) {
        // Check if this line references 'error' but error doesn't exist
        if (line.includes('error instanceof Error') && !line.includes('catch')) {
          // Look for the actual error variable in nearby lines
          let actualErrorVar = null;
          // Check previous 5 lines for catch block
          for (let j = Math.max(0, i - 5); j < i; j++) {
            const prevLine = lines[j];
            const catchMatch = prevLine.match(/catch\s*\(([^:]+)(?::\s*\w+)?\)/);
            if (catchMatch) {
              actualErrorVar = catchMatch[1].trim();
              break;
            }
          }
          
          if (actualErrorVar && actualErrorVar !== 'error') {
            lines[i] = line.replace(/\berror\b/g, actualErrorVar);
            modified = true;
          } else if (!actualErrorVar) {
            // No error variable found, this line is probably dead code, comment it out
            lines[i] = '// ' + line;
            modified = true;
          }
        }
      }
    }
    
    if (modified) {
      content = lines.join('\n');
    }

    if (modified) {
      fs.writeFileSync(filePath, content, 'utf8');
      console.log(`✅ Fixed: ${path.relative(process.cwd(), filePath)}`);
      fixedCount++;
    }
  }

  console.log(`\n📊 Fixed ${fixedCount} files`);
}

fixSyntaxErrors().catch(console.error);

File: scripts/fix-syntax-errors.py
#!/usr/bin/env python3

"""
Script to fix TypeScript syntax errors caused by malformed import/export statements.
This script processes files to fix common syntax patterns that cause compilation errors.
"""

import os
import re
import sys
import shutil
from pathlib import Path
from datetime import datetime
import subprocess

def create_backup(file_path):
    """Create a backup of the file before modifying it."""
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    backup_path = f"{file_path}.backup_{timestamp}"
    shutil.copy2(file_path, backup_path)
    return backup_path

def fix_import_export_syntax(content):
    """Fix common import/export syntax errors."""
    original = content
    
    # Fix pattern: "export { Something, } from './types';"
    content = re.sub(r'export\s*{\s*([^,}]+),\s*}\s*from', r'export { \1 } from', content)
    
    # Fix pattern: "import { Something, } from './types';"
    content = re.sub(r'import\s*{\s*([^,}]+),\s*}\s*from', r'import { \1 } from', content)
    
    # Fix duplicate 'import {' patterns like "import { import {"
    content = re.sub(r'import\s*{\s*import\s*{', 'import {', content)
    content = re.sub(r'export\s*{\s*export\s*{', 'export {', content)
    
    # Fix "export type { import {" pattern
    content = re.sub(r'export\s+type\s*{\s*import\s*{', 'export type {', content)
    
    # Fix malformed multi-line imports/exports with "} from './types';" on wrong line
    content = re.sub(r'([,\s]+)\s*}\s*from\s*[\'"]\.\/types[\'"];', r' } from "./types";', content)
    
    # Fix imports from wrong paths
    content = re.sub(r'from\s*[\'"]\.\.\/types\/canonical[\'"]', "from '../../types/canonical'", content)
    content = re.sub(r'from\s*[\'"]\.\/types[\'"]', "from '../types'", content)
    
    # Fix trailing commas before closing brace
    content = re.sub(r',\s*}\s*from', ' } from', content)
    
    # Fix orphaned "} from" statements
    content = re.sub(r'^\s*}\s*from\s*[\'"][^\'\"]+[\'\"];?\s*$', '', content, flags=re.MULTILINE)
    
    # Fix multiple consecutive commas
    content = re.sub(r',\s*,+', ',', content)
    
    # Fix empty imports/exports
    content = re.sub(r'(import|export)\s*{\s*}\s*from\s*[\'"][^\'\"]+[\'\"];?', '', content)
    
    # Clean up extra whitespace
    content = re.sub(r'[ \t]+', ' ', content)
    content = re.sub(r'\n\s*\n\s*\n+', '\n\n', content)
    
    return content

def fix_mantine_imports(content):
    """Fix imports that incorrectly import from canonical types instead of Mantine."""
    # Components that should come from @mantine/core
    mantine_components = [
        'Table', 'ScrollArea', 'Badge', 'Group', 'Text', 'ActionIcon',
        'Paper', 'Stack', 'Checkbox', 'Button', 'Modal', 'TextInput',
        'Select', 'Textarea', 'Switch', 'Tabs', 'Card', 'Grid', 'Col',
        'Container', 'Center', 'Flex', 'SimpleGrid', 'Space', 'Divider'
    ]
    
    for component in mantine_components:
        # Fix imports that incorrectly import these from canonical
        pattern = rf"import\s*{{\s*([^}}]*\b{component}\b[^}}]*)}}\s*from\s*['\"].*?/types/canonical['\"]"
        
        def replacer(match):
            components = match.group(1)
            # Remove the component from this import
            remaining = re.sub(rf'\b{component}\b,?\s*', '', components).strip().rstrip(',')
            if remaining:
                return f"import {{ {remaining} }} from '../../types/canonical'"
            else:
                return ''
        
        content = re.sub(pattern, replacer, content)
        
        # Add proper Mantine import if component is used but not imported
        if component in content and f"from '@mantine/core'" not in content:
            # Add to existing Mantine import if present
            mantine_import_pattern = r"import\s*{\s*([^}]+)\s*}\s*from\s*['\"]@mantine/core['\"]"
            match = re.search(mantine_import_pattern, content)
            if match:
                existing_imports = match.group(1)
                if component not in existing_imports:
                    new_imports = f"{existing_imports.rstrip()}, {component}"
                    content = re.sub(mantine_import_pattern, 
                                   f"import {{ {new_imports} }} from '@mantine/core'", 
                                   content, count=1)
    
    return content

def process_file(file_path):
    """Process a single file to fix syntax errors."""
    try:
        # Read file
        with open(file_path, 'r', encoding='utf-8') as f:
            content = f.read()
        
        # Create backup
        backup_path = create_backup(file_path)
        
        # Apply fixes
        fixed_content = fix_import_export_syntax(content)
        fixed_content = fix_mantine_imports(fixed_content)
        
        # Write back if changed
        if fixed_content != content:
            with open(file_path, 'w', encoding='utf-8') as f:
                f.write(fixed_content)
            print(f"✅ Fixed: {file_path}")
            return True
        else:
            # Remove backup if no changes
            os.remove(backup_path)
            print(f"⏭️  No changes needed: {file_path}")
            return False
    except Exception as e:
        print(f"❌ Error processing {file_path}: {e}")
        return False

def get_files_with_errors():
    """Get list of files with TypeScript errors."""
    try:
        result = subprocess.run(
            ["pnpm", "type-check"],
            capture_output=True,
            text=True,
            timeout=60
        )
        
        # Parse error output to get file paths
        files = set()
        for line in result.stderr.split('\n'):
            if 'error TS' in line:
                # Extract file path from error line
                parts = line.split('(')
                if parts:
                    file_path = parts[0].strip()
                    if file_path and os.path.exists(file_path):
                        files.add(file_path)
        
        return list(files)
    except Exception as e:
        print(f"Error getting files with errors: {e}")
        return []

def main():
    """Main function to process all files with syntax errors."""
    print("🔍 Finding files with TypeScript errors...")
    
    error_files = get_files_with_errors()
    
    if not error_files:
        print("No files with errors found or unable to get error list.")
        print("Processing known problematic files...")
        
        # Fallback to known problematic files
        error_files = [
            "src/api/metadataProviders/fandomClient.ts",
            "src/components/addManga/MetadataEditor.tsx",
            "src/components/addManga/form.tsx",
            "src/components/addManga/mangaSearchResult.tsx",
            "src/components/addLibrary.tsx",
            "src/components/addManga/components/display/MetadataPreview.tsx",
            "src/components/addManga/components/display/ProviderResults.tsx",
            "src/components/addManga/components/media/ImageGallery.tsx",
            "src/components/addManga/components/media/VolumeCoversGallery.tsx",
            "src/components/addManga/components/media/BannerGallery.tsx",
            "src/components/addManga/components/media/ChapterCoversGallery.tsx",
            "src/components/addManga/components/performance/LazyLoad.tsx",
            "src/components/addManga/components/performance/VirtualList.tsx",
            "src/components/addManga/components/performance/index.ts",
            "src/components/addManga/components/utils/index.ts"
        ]
        
        # Filter to only existing files
        error_files = [f for f in error_files if os.path.exists(f)]
    
    print(f"📝 Found {len(error_files)} files to process")
    
    fixed_count = 0
    for i, file_path in enumerate(error_files, 1):
        print(f"\n[{i}/{len(error_files)}] Processing {file_path}...")
        if process_file(file_path):
            fixed_count += 1
    
    print(f"\n✨ Completed! Fixed {fixed_count} out of {len(error_files)} files")
    
    # Run type check to show improvement
    print("\n📊 Running type check to see improvements...")
    result = subprocess.run(
        ["pnpm", "type-check"],
        capture_output=True,
        text=True
    )
    
    error_count = result.stderr.count('error TS')
    print(f"Current error count: {error_count}")

if __name__ == "__main__":
    main()